content/blog/2025-04-04-datafusion-userdefined-window-functions.md

```diff
@@ -0,0 +1,339 @@
+
+- Alternative approaches and their trade-offs
+
+## Understanding Window Functions in SQL 
+
+Imagine you're analyzing sales data and want insights without losing the finer details. This is where **window functions** come into play. Unlike **GROUP BY**, which condenses data, window functions let you retain each row while performing calculations over a defined **range** —like having a moving lens over your dataset.
+
+Picture a business tracking daily sales. They need a running total to understand cumulative revenue trends without collapsing individual transactions. SQL makes this easy:
+```sql
+SELECT id, value, SUM(value) OVER (ORDER BY id) AS running_total
+FROM sales;
+```
+```text
+example:
++------------+--------+-------------------------------+
+|   Date     | Sales  | Rows Considered               |
++------------+--------+-------------------------------+
+| Jan 01     | 100    | [100]                         |
+| Jan 02     | 120    | [100, 120]                    |
+| Jan 03     | 130    | [100, 120, 130]               |
+| Jan 04     | 150    | [100, 120, 130, 150]          |
+| Jan 05     | 160    | [100, 120, 130, 150, 160]     |
+| Jan 06     | 180    | [100, 120, 130, 150, 160, 180]|
+| Jan 07     | 170    | [100, ..., 170] (7 days)      |
+| Jan 08     | 175    | [120, ..., 175]               |
++------------+--------+-------------------------------+
+A row-by-row representation of how a 7-day moving average includes the previous 6 days and the current one.
+```
+
+
+This helps in analytical queries where we need cumulative sums, moving averages, or ranking without losing individual records.
+
+
+## User Defined Window Functions
+Built-in window functions serve many use cases, but sometimes custom logic is needed—for example:
+
+- Calculating moving averages with complex conditions
+
+- Implementing a custom ranking strategy
+
+- Tracking non-standard cumulative logic
+
+Thus, **User-Defined Window Functions (UDWFs)** allow developers to define their own behavior using a combination of SQL and *bit of logic*.
+
+Writing a user defined window function is slightly more complex than an aggregate function due
+to the variety of ways that window functions are called. I recommend reviewing the
+[online documentation](https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf)
+for a description of which functions need to be implemented. 
+
+## Understaing Sliding Window 
+
+Sliding windows define a **moving range** of data over which aggregations are computed. Unlike simple cumulative functions, these windows are dynamically updated as new data arrives.
+
+For instance, if we want a 7-day moving average of sales:
+
+```sql
+SELECT date, sales, 
+       AVG(sales) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS moving_avg
+FROM sales;
+```
+Here, each row’s result is computed based on the last 7 days, making it computationally intensive as data grows.
```

Comment:
I am not sure I would say a 7 day window is computationally expensive -- if you changed the example to be `UNBOUNDED PRECEDING` you could make the point that the window size kept growing. Or 365 days maybe would make the point better?